src/java/org/apache/cassandra/io/util/CompressedChunkReader.java

```diff
@@ -35,15 +35,22 @@ public abstract class CompressedChunkReader extends AbstractReaderFileProxy impl
 {
     final CompressionMetadata metadata;
     final int maxCompressedLength;
+    protected final long startOffset;
 
-    protected CompressedChunkReader(ChannelProxy channel, CompressionMetadata metadata)
+    protected CompressedChunkReader(ChannelProxy channel, CompressionMetadata metadata, long startOffset)
     {
-        super(channel, metadata.dataLength);
+        super(channel, metadata.dataLength + startOffset);
         this.metadata = metadata;
         this.maxCompressedLength = metadata.maxCompressedLength();
+        this.startOffset = startOffset;
         assert Integer.bitCount(metadata.chunkLength()) == 1; //must be a power of two
     }
 
+    protected long sliceChunkOffset(long originalChunkOffset)
+    {
+        return startOffset == 0 ? originalChunkOffset : originalChunkOffset - metadata.chunkFor(startOffset).offset;
```

Comment:
`metadata.chunkFor(startOffset).offset` involves a binary search. Let's store that in a final field to avoid redoing that search.

Reply:
It does not, for compression metadata the main cost is with creating a `Chunk` object. Anyway, I've already fixed that in the last commit